src/lib/utils/socket/uri.cpp

```diff
@@ -0,0 +1,155 @@
+/*
+* (C) 2019 Nuno Goncalves <nunojpg@gmail.com>
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/uri.h>
+#include <botan/exceptn.h>
+
+#if defined(BOTAN_HAS_BOOST_ASIO)
+   #include <boost/asio.hpp>
+#endif
+
+namespace {
+bool isIPv4(const std::string& ip)
+   {
+#if defined(BOTAN_HAS_BOOST_ASIO)
+   boost::system::error_code ec;
+   boost::asio::ip::address_v4::from_string(ip, ec);
+   if(ec)
+      {
+      return false;
+      }
+#else
+   (void)ip; //squelch unused variable warning
+#endif
+   return true;
+   }
+
+bool isIPv6(const std::string& ip)
+   {
+#if defined(BOTAN_HAS_BOOST_ASIO)
+   boost::system::error_code ec;
+   boost::asio::ip::address_v6::from_string(ip, ec);
+   if(ec)
+      {
+      return false;
+      }
+#else
+   (void)ip; //squelch unused variable warning
+#endif
+   return true;
+   }
+}
+
+namespace Botan {
+
+URI URI::fromDomain(const std::string& uri) noexcept
+   {
+   unsigned port = 0;
+   const auto port_pos = uri.find(':');
+   if(port_pos != std::string::npos)
+      {
+      for(char c : uri.substr(port_pos+1))
+         {
+         if(!std::isdigit(c))
+            return {};
+         port = port*10 + c - '0';
+         if(port > 65535)
+            return {};
```

Comment:
Would it be a bad idea if it was an exception thrown instead for all the `from...` ?

Reply:
Done.